### ISSUE TITLE
Certain MythicMobs asren't being recognised in /aa, or, possible permission handling not working right

### ISSUE THREAD

**Issue by DXUltimate** [PyvesB/advanced-achievements#831]
:information_source: **Basic information**

* Advanced Achievements version: 6.2
* Minecraft version: 1.16.1
* [Wiki](https://github.com/PyvesB/advanced-achievements/wiki) read: Yes
* [Existing issues](https://github.com/PyvesB/advanced-achievements/issues) searched: Couldn't really find anything retaining this issue I'm about to post, no.

:beetle: **Description**

So, we have multiple categories for our Kill achievements, including MM mobs. Here are the achieves;

Rare Mobs
Elite Mobs
Chnampion Mobs
Giants
Mini-Bosses
Bosses

So the Rare mobs tier of achieves completely work, but the Elites, and giants do not. Elite achievements only work for op'd players, despite having the permission node listed in the group.

https://puu.sh/GlQj0/ad4b35b0ef.png

(green means true, while red stands for false)

:page_facing_up: **Server logs**

No logs believe it or not, even upon loading/restarting. =)

:wrench: **Configuration**

````
config.yml
https://pastebin.com/N5BXBEq2

MM Config for the Elites
https://pastebin.com/Tquzw4rM

Any more info can be given on a moments notice. =) Thanks for all your work on this amazing plugin.


````

**Comment by PyvesB** [comment]
Hello @DXUltimate,

Thanks for sharing all the details. Could you please clarify what exactly isn't working: are the statistics not increasing at all or is it just the final achievements that aren't being awarded?

**Comment by DXUltimate** [comment]
Statistics aren't increasing at all when players are actually fulfilling
said requirements. How we have it set up is we have mob arenas, linked to
crates, which people use "keys" to get mobs, these Elite Skeletons for
example with the grey text. It was a bit confusing at first as "op'd"
players can get the Elites, and only 2 of the 4 giants we have listed in
the achievements, while any other players on the server, cannot, even if
they have the permission node listed for the achievement/category.

Speaking of the giant's section, another strange thing is this.
https://puu.sh/Gn4K0/6b75b1f2d9.png - This image shows that I received the
first Giant tier achievement, however
https://puu.sh/Gn4IW/422e983ea5.png - This image shows that total progress
the the rest of them, is 0, however, should infact be "1" due to the above
image.

Lastly, I created a /spawn achievement tier list in the commands section,
all it is, is just doing /spawn 1000 times, however, our users and ops are
unable to get this achievement as well.

Now, I know it is possible to reset the counters via commands, but I can
assure you that I have not used any of the commands to reset any of the
progress of the counters. I did use one for the achievements, but I was
doing that to test Fishing, as we use a custom enchantments plugin, which
allowed for instant cooked fish.

Thanks for your time on this. If you require additional information, I will
be happy to assist. =)

**Comment by PyvesB** [comment]
I'm unable to spot the origin of the permissions problem by looking at the configs, so I've prepared a slightly modified Advanced Achievements version for you, which may help with debugging:
[AdvancedAchievements.zip](https://github.com/PyvesB/advanced-achievements/files/5154453/AdvancedAchievements.zip)

It contains additional log statements, containing the keyword with `[PERMISSIONS]` so that you can easily spot them:
* log indicating which permissions are dynamically being added by the plugin at startup.
* log indicating which permission is being checked when an achievement is received (though based on your previous message this is probably not a problem).
* log indicating which permissions are being checked when an entity is killed.

This will add quite a lot of logs to your server, but I wouldn't be too worried about performance. You can swap out the version as soon as you've collected enough information.

What I'm hoping is that we'll see the plugin checking for a permission (bullet point 3) that isn't set on startup (bullet point 1) or that you haven't configured properly in your permissions plugin. :wink:

>  Now, I know it is possible to reset the counters via commands, but I can assure you that I have not used any of the commands to reset any of the progress of the counters. I did use one for the achievements, but I was doing that to test Fishing, as we use a custom enchantments plugin, which allowed for instant cooked fish.

Maybe you've added new mobs to the grouped category?

**Comment by PyvesB** [comment]
@DXUltimate is your issue solved? If not, please check the previous message to get more logs. :wink:

**Comment by DXUltimate** [comment]
I've been very busy looking for work and whatnot, so I apologize. I will do the above and get back to you within a few hours. 

Thanks again. I am SSS.BarX on spigot btw, if you needed that for verification purposes. =)

**Comment by DXUltimate** [comment]
Okay here are my findings.

Permission plugin I use is LuckPerms 5.1.13

First of all, this version definitely helped with my issue, so thank you heaps for that. The specific permission node that group "default" was missing was "achievement.count.kills.*" which allowed the Elite mobs listed above to be killed, no luck with the giants with this, though. Console did pop up with a message saying;

[04:29:06] [Server thread/INFO]: [AdvancedAchievements] [PERMISSIONS] Checking permission achievement.count.kills.giant for Kills event.
[04:29:06] [Server thread/INFO]: [AdvancedAchievements] [PERMISSIONS] Checking permission achievement.count.kills.Angry Giant for Kills event.
[04:29:06] [Server thread/INFO]: CONSOLE issued server command: /broadcast Iji3 killed the Angry Giant

After that, I tried the node "achievement.count.* and the "achievement.count.kills", as well as both of them at the exact same time, Iji3 did get credited for the Angry Giant kill. I feel like it was the achievement.count that initially game iji the ability to gain it, but I will have to do further testing.

TL;DR ver, for whatever reason permission node achievement.count.kills.* did not work, but achievement.count.* did (both both at once did, which is very strange, if you think about it), and allowed the kills to happen.

I am going to continue using this version of AA to diagnose /spawn achievement, as that is not working either, but for now, thank you very much for this. <3

**Comment by PyvesB** [comment]
Thanks, that's useful information. One of the problems is that `achievement.count.kill.*` is not a permission available in the plugin. This is admittedly confusing, so will fix that. I'll also add some sort of log message or _/aach perms_ command to list all permissions registered by Advanced Achievements, that will help people without having to build custom plugin builds which soon become outdated. :)

**Comment by PyvesB** [comment]
I spotted another problem in the code, the plugin should remove spaces from the custom mob's name when checking for permissions, e.g. `achievement.count.kills.Angry` Giant -> `achievement.count.kills.AngryGiant`. That is now fixed.

I'll drop the command idea: even for servers with not that many achievements, there are simply too many permissions nodes to be displayed in the in-game chat or the console. Some servers could easily end up with several hundreds or thousand lines... Another problem is that permissions can have color sequences in them, e.g. `achievement.count.kills.§2SkeletalKnight`, those get messed up when displayed in any Minecraft console/chat. I'll simply polish documentation instead, that will hopefully be enough for most users.

All issues in the plugin should have been addressed, so I'll go ahead and close this. Feel free to reopen or open a new issue if you are still facing problems when the new version gets released! :)

**Comment by DXUltimate** [comment]
<3

Thanks for all your work man. You have our server in full support forever. =)

-DX
